### DISCUSSION TITLE
[Improvement]: support multiple composer.lock files

### DISCUSSION THREAD

**Discussion by passioneight** [pimcore/pimcore#12441] [General]
### Improvement description

# Description
It would be nice to be able to define the `composer.lock` file that is used for loading the available composer bundles, used here:
https://github.com/pimcore/pimcore/blob/a020f961b19ead3899215f8c9188e763e9741536/lib/Composer/PackageInfo.php#L62

Preferably, as option for the `bin/console pimcore:bundle:enable` command - unless, anyone has a better idea. 

# Details
Developing locally, our setup requires multiple composer files:
1. `composer.json`, which will require any _released_ bundle, so it is available in our developer-sandbox to interactively showcase what they are used for and how to use them.
2. `composer.dev.local.json`, which is used to add repositories as `@dev`, so they are symlinked by composer (using the `repositories` section) for local development.

However, one cannot enable the bundle using `bin/console pimcore:bundle:enable <bundle-name>`, because only the `composer.lock` file is considered, which does not contain the bundle just yet.

**Comment by dpfaffenbauer** [comment]
how do you differentiate between composer.dev.local.json and composer.json? As far as composer is concerned, that is not a "standard" use-case.

**Comment by passioneight** [comment]
You can add `COMPOSER=composer.dev.local.json` in front of your composer command to use a different file.

The official composer repository is actually were I got the idea from: someone else wanted pretty much the same setup, so the guys from composer suggested this approach.

I agree, though, it doesn't really seem to be a common use case. Would you like to share your approach on bundle development across your team?

**Comment by dpfaffenbauer** [comment]
We either develop the bundle directly inside the project using local directories with composer:

```
"repositories": {
  "some-name": {
      "type": "path",
      "url": "lib/Path/To/Bundle"
    }
"
```

Or we develop the bundle completely independent with tests and stuff and include it in the project like a regular dependency.

**Comment by passioneight** [comment]
I just had another look at this and am wondering if @brusch knows why the `composer.lock` is used at all, when the `installed.json` - which is used as fallback - contains the _actually_ installed packages anyway.

### Update
Using only the following lines, everything works as expected:
https://github.com/pimcore/pimcore/blob/a020f961b19ead3899215f8c9188e763e9741536/lib/Composer/PackageInfo.php#L68-L71

**Reply by jdreesen** [reply]
You're right, the `vendor/composer/installed.json` does seem to contain everything the `composer.lock` does (with even three additional keys `version_normalized`, `installation-source` and `install-path`).

The question is, what was the reason to prefer reading the `composer.lock` file? But it seems that we don't need it and can rely on the `vendor/composer/installed.json`.

**Reply by passioneight** [reply]
Seems like this was added by @maff quite a while ago: https://github.com/pimcore/pimcore/commit/05e557b56ae05954f1c94b82955bcd6681193258#diff-7ca9cbdd89e8cf55b072879cd4e645a11f9f2be07526cd65f41c0268c03a3ebaR128

I imagine this to be legacy code and it can therefore be removed, wdyt?

**Reply by jdreesen** [reply]
Yes, it's there since the beginning, which makes it hard to tell where it's exactly needed for. Sadly, @maff isn't here anymore to tell us why it was added in the first place.

I'd say we don't need it, and it can be removed.

**Reply by passioneight** [reply]
Alright then, [here's the new issue](https://github.com/pimcore/pimcore/issues/12491).

**Comment by passioneight** [comment]
@dpfaffenbauer thank you for sharing - much appreciated!

We too developed bundles within the projects, but that lead to many accidental dependencies to the `App` namespace within the bundle. 

The second approach is pretty much what I'm trying to achieve. Only that the bundle is supposed to be in another, internal project, which can be deployed whenever a new feature was implemented. This way, we can directly showcase new features and raise awareness of any available functionality.
